src/base/utils/fs.cpp

```diff
@@ -128,18 +132,25 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
               , [](const QString &l, const QString &r) { return l.count('/') > r.count('/'); });
 
     for (const QString &p : asConst(dirList)) {
-        // remove unwanted files
-        for (const QString &f : deleteFilesList) {
-            forceRemove(p + f);
-        }
-
-        // remove temp files on linux (file ends with '~'), e.g. `filename~`
         const QDir dir(p);
+        // A deeper folder may have not been removed in the previous iteration
+        // so don't remove anything from this folder either.
+        if (!dir.isEmpty(QDir::AllDirs | QDir::NoDotAndDotDot))
```

Comment:
I overlooked this, wouldn't this be slow since it is going to enumerate all the subfolders every time and count them?
https://doc.qt.io/qt-5/qdir.html#isEmpty

This should overcome that problem:
```c++
bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
{
    /* same code, comment them out
    if (path.isEmpty() || !QDir(path).exists())
        return true;

    const QStringList deleteFilesList = {
        // Windows
        "Thumbs.db",
        "desktop.ini",
        // Linux
        ".directory",
        // Mac OS
        ".DS_Store"
    };

    // travel from the deepest folder and remove anything unwanted on the way out.
    QStringList dirList(path + '/');  // get all sub directories paths
    QDirIterator iter(path, (QDir::AllDirs | QDir::NoDotAndDotDot), QDirIterator::Subdirectories);
    while (iter.hasNext())
        dirList << iter.next() + '/';
    // sort descending by directory depth
    std::sort(dirList.begin(), dirList.end()
              , [](const QString &l, const QString &r) { return l.count('/') > r.count('/'); });
   */

    const bool foundOtherFiles = std::any_of(dirList.cbegin(), dirList.cend(), [&deleteFilesList](const QString &path) {
        const QDir dir(path);
        const QStringList tmpFileList = dir.entryList(QDir::Files);
        return std::any_of(tmpFileList.cbegin(), tmpFileList.cend(), [&deleteFilesList](const QString &f)
        {
            return (!f.endsWith('~') && !deleteFilesList.contains(f, Qt::CaseInsensitive));
        });
    });
    if (foundOtherFiles)
        return false;

    removeDirRecursive(path);
    return QDir(path).exists();
}
```

Reply:
Aren't you a bit obsessed on not introducing slowness, even in seldomly called functions?
In any case the docs state:
>Equivalent to count() == 0 with filters QDir::AllEntries | QDir::NoDotAndDotDot, **but faster as it just checks whether the directory contains at least one entry.**

I am worried if I should have used `QDir::Dirs` instead. I don't understand the difference between that and `QDir::AllDirs` from the docs.

PS: I can't clearly follow your proposed approached. I don't know if it can account for a branch in the subfolders where one branch isn't empty and the other is. Will it remove the empty branch up to the point where the 2 branches have the common parent-folder?

Reply:
or disregard my comment/code above (not good as now I rethink) and use `if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot)) return false;`

Reply:
You mean:
```c++
 if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot))
    continue;
```
right?

Reply:
>I am worried if I should have used QDir::Dirs instead. I don't understand the difference between that and QDir::AllDirs from the docs.

Yes, it seems `QDir::AllDirs` has special effects when using `QDir::entryList()`. I think most other time you should use `QDir::Dirs`.

>Aren't you a bit obsessed on not introducing slowness, even in seldomly called functions?

Sorry, but from algorithm viewpoint the current code feels uneasy for me. A naive example is querying over a network drive with caching disabled, then practically you'll want to avoid any redundant operation.
BTW, IMO performance is always a consideration when writing code and programmer should not easily neglect/trade off performance for nothing.
Taking quote from https://ubiquity.acm.org/article.cfm?id=1513451:
>But, conversely, when designing software at a system level, performance issues should always be considered from the beginning. A good software developer will do this automatically, having developed a feel for where performance issues will cause problems. An inexperienced developer will not bother, misguidedly believing that a bit of fine tuning at a later stage will fix any problems.

Reply:
> You mean:
> 
> ```c++
>  if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot))
>     continue;
> ```
> 
> right?

No. Can't you use `return false;` here?

Reply:
>Not exactly. Can't you use return false; here?

We're inside a loop of folders. Consider this folder tree
>a->b->ca->d->3rd_party_file
>a->b->cb

And smart remove is called on `a`. In this case `cb` should be removed but not `b->ca->d->3rd_party_file`

Reply:
>And smart remove is called on a. In this case cb should be removed but not b->ca->d->3rd_party_file

OK for me if that is intended.

@glassez 
Just wanted to notify you because now it seems a bit different to what you said at: https://github.com/qbittorrent/qBittorrent/pull/11660#issuecomment-565709115

Reply:
I meant the files. Deleting an empty subfolder probably shouldn't be a problem. But I do not have any personal preferences in this problem, because I prefer not to interfere with automated processes.